src/lazy/encoder/write_as_ion.rs

```diff
@@ -296,29 +296,41 @@ impl<'a, D: Decoder> WriteAsIon for ValueRef<'a, D> {
             String(s) => value_writer.write_string(s.text()),
             Clob(c) => value_writer.write_clob(c.as_ref()),
             Blob(b) => value_writer.write_blob(b.as_ref()),
-            List(l) => {
-                let mut list = value_writer.list_writer()?;
-                for value in l {
-                    list.write(value?.read()?)?;
-                }
-                list.close()
-            }
-            SExp(s) => {
-                let mut sexp = value_writer.list_writer()?;
-                for value in s {
-                    sexp.write(value?.read()?)?;
-                }
-                sexp.close()
-            }
-            Struct(s) => {
-                let mut struct_ = value_writer.struct_writer()?;
-                for field_result in s {
-                    let field = field_result?;
-                    struct_.write(field.name()?, field.value().read()?)?;
-                }
-                struct_.close()
-            }
+            List(l) => value_writer.write(l),
+            SExp(s) => value_writer.write(s),
+            Struct(s) => value_writer.write(s),
```

Comment:
🗺️ Giving these types their own `WriteAsIon` implementations streamlined this code path nicely, but my real aim was to make passing them to a writer in other contexts easier.